File: order-ms/src/main/java/com/nexeum/orderms/service/OrderServiceImpl.java
package com.nexeum.orderms.service;

import com.nexeum.orderms.dto.request.CancelOrderRequest;
import com.nexeum.orderms.dto.request.CreateOrderRequest;
import com.nexeum.orderms.dto.request.GetAllOrderRequest;
import com.nexeum.orderms.dto.response.GetAllOrderResponse;
import com.nexeum.orderms.dto.response.ProductItem;
import com.nexeum.orderms.dto.response.ServiceResponse;
import com.nexeum.orderms.dto.response.Order;
import com.nexeum.orderms.entity.OrderItem;
import com.nexeum.orderms.entity.ProductOrder;
import com.nexeum.orderms.repository.ProductOrderRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
@Slf4j
public class OrderServiceImpl implements OrderService {
    @Autowired
    private ProductOrderRepository productOrderRepository;

    @Override
    public ResponseEntity<?> createOrder(CreateOrderRequest request) {
        ResponseEntity<?> responseEntity;
        ServiceResponse serviceResponse = new ServiceResponse();

        try {
            ProductOrder productOrder = new ProductOrder();
            productOrder.setCartId(request.getCartId());
            productOrder.setOrderStatus("In Progress");

            List<OrderItem> orderItems = new ArrayList<>();

            request.getOrderItems().forEach(orderProductItem -> {
                OrderItem orderItem = new OrderItem();
                orderItem.setProductId(orderProductItem.getProductId());
                orderItem.setProductName(orderProductItem.getProductName());
                orderItem.setProductQuantity(orderProductItem.getProductQuantity());
                orderItem.setUnitPrice(orderProductItem.getUnitPrice());
                orderItem.setTotalPrice(orderProductItem.getTotalPrice());
                orderItem.setProductImageUrl(orderProductItem.getProductImageUrl());
                orderItems.add(orderItem);
            });

            productOrder.setOrderItems(orderItems);
            productOrder.setTotalOrderPrice(request.getTotalOrderPrice());
            productOrderRepository.save(productOrder);
            serviceResponse.setCode("200");
            serviceResponse.setMessage("Order Created Successfully");
            responseEntity = new ResponseEntity<>(serviceResponse, HttpStatus.OK);
        } catch (Exception e) {
            serviceResponse.setCode("500");
            serviceResponse.setMessage("Internal Server Error");
            responseEntity = new ResponseEntity<>(serviceResponse, HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return responseEntity;
    }

    @Override
    public ResponseEntity<?> cancelOrder(CancelOrderRequest request) {
        ResponseEntity<?> responseEntity;
        ServiceResponse serviceResponse = new ServiceResponse();

        try {
            ProductOrder productOrder = productOrderRepository.findById(request.getOrderId()).orElse(null);
            productOrder.setOrderStatus("Cancelled");
            productOrderRepository.save(productOrder);
            serviceResponse.setCode("200");
            serviceResponse.setMessage("Order Cancelled Successfully");
            responseEntity = new ResponseEntity<>(serviceResponse, HttpStatus.OK);
        } catch (Exception e) {
            serviceResponse.setCode("500");
            serviceResponse.setMessage("Internal Server Error");
            responseEntity = new ResponseEntity<>(serviceResponse, HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return responseEntity;
    }

    @Override
    public ResponseEntity<?> getAllOrders(GetAllOrderRequest request) {
        ResponseEntity<?> responseEntity;

        try {
            List<ProductOrder> productOrders = productOrderRepository.findByCartId(request.getCartId());

            if (!(productOrders.isEmpty())) {
                GetAllOrderResponse response = new GetAllOrderResponse();

                List<Order> orders = new ArrayList<>();

                for (ProductOrder productOrder : productOrders) {
                    Order order = new Order();
                    order.setOrderId(productOrder.getId());
                    order.setTotalOrderPrice(productOrder.getTotalOrderPrice());
                    order.setOrderStatus(productOrder.getOrderStatus());
                    order.setTotalOrderPrice(productOrder.getTotalOrderPrice());
                    order.setCreatedDate(productOrder.getCreatedDate());
                    order.setUpdatedDate(productOrder.getUpdatedDate());

                    List<ProductItem> products = new ArrayList<>();

                    productOrder.getOrderItems().forEach(orderItem -> {
                        ProductItem productItem = new ProductItem();
                        productItem.setId(orderItem.getId());
                        productItem.setProductId(orderItem.getProductId());
                        productItem.setProductName(orderItem.getProductName());
                        productItem.setProductQuantity(orderItem.getProductQuantity());
                        productItem.setUnitPrice(orderItem.getUnitPrice());
                        productItem.setTotalPrice(orderItem.getTotalPrice());
                        productItem.setProductImageUrl(orderItem.getProductImageUrl());
                        products.add(productItem);
                    });

                    order.setOrderItems(products);
                    orders.add(order);
                }
                response.setOrders(orders);
                responseEntity = new ResponseEntity<>(response, HttpStatus.OK);
            } else {
                responseEntity = new ResponseEntity<>(HttpStatus.NO_CONTENT);
            }

        }catch (Exception e) {
            responseEntity = new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return responseEntity;
    }
}

File: order-ms/service/OrderServiceImpl.cs
using Microsoft.AspNetCore.Mvc;
using order_ms.dto.request;
using order_ms.dto.response;
using order_ms.models;
using order_ms.repository;

namespace order_ms.service
{
    public class OrderService : IOrderService
    {
        private readonly ProductOrderRepository _productOrderRepository;

        private readonly ILogger<OrderService> _logger;

        public OrderService(ProductOrderRepository productOrderRepository, ILogger<OrderService> logger)
        {
            _productOrderRepository = productOrderRepository;
            _logger = logger;
        }
        public async Task<IActionResult> CreateOrder(CreateOrderRequest request)
        {
            var serviceResponse = new ServiceResponse();

            try
            {
                var productOrder = new ProductOrder
                {
                    CartId = request.CartId.ToString(),
                    OrderStatus = "In progress",
                    OrderItems = (request.OrderItems ?? new List<OrderProductItem>()).Select(oi => new OrderItem
                    {
                        ProductId = oi.ProductId,
                        ProductName = oi.ProductName,
                        ProductQuantity = oi.ProductQuantity,
                        PricePerUnit = oi.PricePerUnit,
                        ProductImageUrl = oi.ProductImageUrl,
                        TotalPrice = oi.TotalPrice
                    }).ToList(),
                    TotalOrderPrice = request.TotalOrderPrice
                };

                await _productOrderRepository.InsertOrderAsync(productOrder);

                serviceResponse.Code = "200";
                serviceResponse.Message = "Order Placed Successfully";

                return new OkObjectResult(serviceResponse);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error occurred while cancelling the order");

                serviceResponse.Code = "500";
                serviceResponse.Message = "Internal Server Error";

                return new OkObjectResult(serviceResponse);
            }
        }
        
        public async Task<IActionResult> CancelOrder(CancelOrderRequest request)
        {
            var serviceResponse = new ServiceResponse();

            try
            {
                var productOrder = await _productOrderRepository.FindByIdAsync(request.OrderId);

                if (productOrder == null)
                {
                    serviceResponse.Code = "404";
                    serviceResponse.Message = "Order not found";
                    return new NotFoundObjectResult(serviceResponse);
                }

                productOrder.OrderStatus = "Cancelled";

                await _productOrderRepository.UpdateOrderAsync(productOrder);

                serviceResponse.Code = "200";
                serviceResponse.Message = "Order Cancelled";

                return new OkObjectResult(serviceResponse);
            }
            catch (Exception)
            {
                serviceResponse.Code = "500";
                serviceResponse.Message = "Internal Server Error";
                
                return new StatusCodeResult(500);
            }
        }

        public async Task<IActionResult> GetAllOrders()
        {
            try
            {
                var allOrders = await _productOrderRepository.GetAllOrdersAsync();

                if (allOrders.Any())
                {
                    var response = new GetAllOrderResponse
                    {
                        Orders = allOrders.Select(o => new Order
                        {
                            OrderId = o.Id,
                            TotalOrderPrice = o.TotalOrderPrice,
                            OrderStatus = o.OrderStatus,
                            CreatedAt = o.CreatedAt,
                            UpdatedAt = o.UpdatedAt,
                            OrderItems = o.OrderItems.Select(oi => new ProductItem
                            {
                                Id = oi.Id,
                                ProductId = oi.ProductId,
                                ProductName = oi.ProductName,
                                ProductQuantity = oi.ProductQuantity,
                                TotalPrice = oi.TotalPrice,
                                PricePerUnit = oi.PricePerUnit,
                                ProductImageUrl = oi.ProductImageUrl
                            }).ToList()
                        }).ToList()
                    };

                    return new OkObjectResult(response);
                }

                return new OkObjectResult(new List<Order>());
            }
            catch (Exception)
            {
                return new StatusCodeResult(500);
            }
        }
    }
}